auth-jwt-service/src/main/java/io/stargate/auth/jwt/AuthJwtService.java

```diff
@@ -69,6 +79,148 @@ public StoredCredentials validateToken(String token) throws UnauthorizedExceptio
+
+    return result;
+  }
+
+  /**
+   * TODO
+   *
+   * @param action
+   * @param token
+   * @param primaryKeyValues
+   * @param tableMetadata
+   * @return
+   * @throws Exception
+   */
+  @Override
+  public ResultSet executeDataWriteWithAuthorization(
+      Callable<ResultSet> action, String token, List<String> primaryKeyValues, Table tableMetadata)
+      throws Exception {
+    JSONObject stargateClaims = extractClaimsFromJWT(token);
+
+    preCheckDataReadWrite(stargateClaims, primaryKeyValues, tableMetadata);
+
+    // Just return the result. No value in doing a post check since we can't roll back anyway.
+    return action.call();
+  }
+
+  @Override
+  public ResultSet executeSchemaReadWithAuthorization(
+      Callable<ResultSet> action, String token, String keyspace, String table) throws Exception {
+    JSONObject stargateClaims = extractClaimsFromJWT(token);
+
+    preCheckSchemaRead(stargateClaims, keyspace, table);
+
+    // TODO: [doug] 2020-11-2, Mon, 17:23 Finish implementing
+    return action.call();
+  }
+
+  @Override
+  public ResultSet executeSchemaWriteWithAuthorization(
+      Callable<ResultSet> action, String token, String keyspace, String table) throws Exception {
+    JSONObject stargateClaims = extractClaimsFromJWT(token);
+
+    preCheckSchemaWrite(stargateClaims, keyspace, table);
+
+    // Just return the result. No value in doing a post check since we can't roll back anyway.
+    return action.call();
+  }
+
+  private JSONObject extractClaimsFromJWT(String token) throws JSONException {
+    // Grab the custom claims from the JWT. It's safe to work with the JWT as a plain Base64 encoded
+    // json object here since by this point we've already authenticated the request.
+    String[] parts = token.split("\\.");
+    String decodedPayload = new String(Base64.getUrlDecoder().decode(parts[1]));
+    JSONObject payload = new JSONObject(decodedPayload);
+    return payload.getJSONObject(CLAIMS_FIELD);
+  }
+
+  private void preCheckSchemaRead(JSONObject stargateClaims, String keyspace, String table) {
+    // TODO: [doug] 2020-11-2, Mon, 17:23 Finish implementing
+  }
+
+  private void preCheckSchemaWrite(JSONObject stargateClaims, String keyspace, String table) {
+    // TODO: [doug] 2020-11-2, Mon, 17:23 Finish implementing
+  }
+
+  private void preCheckDataReadWrite(
+      JSONObject stargateClaims, List<String> primaryKeyValues, Table tableMetadata)
+      throws JSONException, UnauthorizedException {
+    List<Column> keys = tableMetadata.primaryKeyColumns();
+
+    if (primaryKeyValues.size() > keys.size()) {
+      throw new IllegalArgumentException("Provided more primary key values than exists");
+    }
+
+    for (int i = 0; i < primaryKeyValues.size(); i++) {
+      // If one of the columns exist as a field in the JWT claims and the values do not match then
+      // the request is not allowed.
+      if (stargateClaims.has(STARGATE_PREFIX + keys.get(i).name())) {
```

Comment:
Do we use column names without their containing table name? Is this descriptive enough?

Reply:
Whatever field is being used should ideally be a unique identifier for a user. Meaning whether they're querying `shopping_cart` or `orders` they should still be constrained to only seeing their data.

Reply:
Ah, so it's just about finding / checking the user :zap: it was not apparent to me before :)

Are we going to manage JWT for end users of the application that connects to Stargate?

Reply:
No there will need to be doc around this. The idea is use whatever auth provider (e.g. auth0, keycloak, ...) and as long as your JWTs have the proper claims and you pass stargate the URL for JWK (a standard endpoint where auth providers expose the public key for verifying a JWT) then we're able to validate them.

Reply:
Thanks for the info, @dougwettlaufer ! This is LGTM. However, given a very specific nature of the key column links to JWT and the "authz" service, I believe this adds weight to having separate interfaces for the old `AuthenticationService` and the new "authz" methods. WDYT?

Reply:
Sure, I can split into Authentication.java and Authorization.java. I was on the fence about it anyway.